core/src/sigverifier/bls_sigverifier.rs

```diff
@@ -422,4 +720,361 @@ mod tests {
+            packet.populate_packet(None, &consensus_message).unwrap();
+            packets.push(packet);
+        }
+
+        let packet_batches = vec![PinnedPacketBatch::new(packets).into()];
+        let result_batches = verifier.verify_batches(packet_batches, num_votes);
+
+        let mut processed_packets = 0;
+        for batch in result_batches {
+            for (i, packet) in batch.iter().enumerate() {
+                processed_packets += 1;
+                if i as u16 == invalid_rank {
+                    assert!(
+                        packet.meta().discard(),
+                        "Packet at invalid rank {i} should be discarded",
+                    );
+                } else {
+                    assert!(
+                        !packet.meta().discard(),
+                        "Packet at valid rank {i} should not be discarded",
+                    );
+                }
+            }
+        }
+        assert_eq!(processed_packets, num_votes, "Did not process all packets");
+    }
+
+    #[test]
+    fn test_blssigverifier_verify_votes_empty_batch() {
+        let (verified_vote_sender, _) = crossbeam_channel::unbounded();
+        let (message_sender, _) = crossbeam_channel::unbounded();
+        let (_, verifier) =
+            create_keypairs_and_bls_sig_verifier(verified_vote_sender, message_sender);
+
+        let packet_batches: Vec<PacketBatch> = vec![];
+        let result_batches = verifier.verify_batches(packet_batches, 0);
+
+        assert!(
+            result_batches.is_empty(),
+            "Result should be an empty vec for empty input"
+        );
+    }
+
+    #[test]
+    fn test_verify_certificate_base2_valid() {
+        let (verified_vote_sender, _) = crossbeam_channel::unbounded();
+        let (message_sender, _) = crossbeam_channel::unbounded();
+        let (validator_keypairs, verifier) =
+            create_keypairs_and_bls_sig_verifier(verified_vote_sender, message_sender);
+
+        let num_signers = 7; // > 60% of 10 validators
+        let slot = 10;
+        let block_hash = Hash::new_unique();
+
+        let certificate = Certificate::Notarize(slot, block_hash);
+
+        let original_vote = Vote::new_notarization_vote(slot, block_hash);
+        let signed_payload = bincode::serialize(&original_vote).unwrap();
+
+        let signatures: Vec<SignatureProjective> = (0..num_signers)
+            .map(|i| validator_keypairs[i].bls_keypair.sign(&signed_payload))
+            .collect();
+        let signature_refs: Vec<&SignatureProjective> = signatures.iter().collect();
+        let aggregate_signature: BlsSignature = SignatureProjective::aggregate(&signature_refs)
+            .unwrap()
+            .into();
+
+        let mut bitmap = BitVec::<u8, Lsb0>::new();
+        bitmap.resize(num_signers, false);
+        for i in 0..num_signers {
+            bitmap.set(i, true);
+        }
+        let encoded_bitmap = encode_base2(&bitmap).unwrap();
```

Comment:
nit: can we use `vote_certificate_builder` so we do end-to-end verification here?

Reply:
Done!